ibm/resource_ibm_network_gateway.go

```diff
@@ -0,0 +1,996 @@
+		filter.Build(
+			filter.Path("backendNetworkComponents.status").Eq("ACTIVE"),
+		),
+	).Id(id).GetBackendNetworkComponents()
+
+	if err != nil {
+		return fmt.Errorf("Error retrieving bare metal server network: %s", err)
+	}
+
+	if len(backendNetworkComponent) > 2 && result.PrimaryBackendNetworkComponent != nil {
+		if *result.PrimaryBackendNetworkComponent.RedundancyEnabledFlag {
+			d.Set("redundant_network", true)
+		} else {
+			d.Set("unbonded_network", true)
+		}
+	}
+
+	if result.OperatingSystem != nil &&
+		result.OperatingSystem.SoftwareLicense != nil &&
+		result.OperatingSystem.SoftwareLicense.SoftwareDescription != nil &&
+		result.OperatingSystem.SoftwareLicense.SoftwareDescription.ReferenceCode != nil {
+		d.Set("os_reference_code", *result.OperatingSystem.SoftwareLicense.SoftwareDescription.ReferenceCode)
+	}
+
+	tagReferences := result.TagReferences
+	tagReferencesLen := len(tagReferences)
+	if tagReferencesLen > 0 {
+		tags := make([]string, 0, tagReferencesLen)
+		for _, tagRef := range tagReferences {
+			tags = append(tags, *tagRef.Tag.Name)
+		}
+		d.Set("tags", tags)
+	}
+
+	storages := result.AllowedNetworkStorage
+	if len(storages) > 0 {
+		d.Set("block_storage_ids", flattenBlockStorageID(storages))
+		d.Set("file_storage_ids", flattenFileStorageID(storages))
+	}
+
+	connInfo := map[string]string{"type": "ssh"}
+	if !*result.PrivateNetworkOnlyFlag && result.PrimaryIpAddress != nil {
+		connInfo["host"] = *result.PrimaryIpAddress
+	} else {
+		connInfo["host"] = *result.PrimaryBackendIpAddress
+	}
+	d.SetConnInfo(connInfo)
+
+	return nil
+}
+func resourceIBMNetworkGatewayUpdate(d *schema.ResourceData, meta interface{}) error {
+	id, _ := strconv.Atoi(d.Id())
+	service := services.GetHardwareService(meta.(ClientSession).SoftLayerSession())
+
+	if d.HasChange("tags") {
+		err := setHardwareTags(id, d, meta)
+		if err != nil {
+			return err
+		}
+	}
+
+	if d.HasChange("notes") {
+		err := setHardwareNotes(id, d, meta)
+		if err != nil {
+			return err
+		}
+	}
+	err := modifyStorageAccess(service.Id(id), id, meta, d)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceIBMNetworkGatewayDelete(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	service := services.GetHardwareService(sess)
+
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return fmt.Errorf("Not a valid ID, must be an integer: %s", err)
+	}
+
+	_, err = waitForNoBareMetalActiveTransactions(id, meta)
+	if err != nil {
+		return fmt.Errorf("Error deleting Network Gateway while waiting for zero active transactions: %s", err)
+	}
+
+	billingItem, err := service.Id(id).GetBillingItem()
+	if err != nil {
+		return fmt.Errorf("Error getting billing item for Network Gateway: %s", err)
+	}
+
+	// Monthly  Softlayer items only support an anniversary date cancellation option.
+	billingItemService := services.GetBillingItemService(sess)
+	_, err = billingItemService.Id(*billingItem.Id).CancelItem(
```

Comment:
Is it possible for billingItem to be nil if some one cancelled the device outside terraform? In that case `*billingItem.Id` will blow. I see a similar problem in `ibm_bare_metal`. We should validate this and check for nilness if that is the case.

Reply:
I added this code:
	billingItem, err := service.Id(id).GetBillingItem()
	if err != nil || *billingItem.Id == 0 {
		return fmt.Errorf("Error getting billing item for Network Gateway: %s", err)
	}
Where assuming that billingItem.Id would get a zero value in case that the item has been deleted

Reply:
I think the check should be more for `billingItem` to be not nil then `*billingItem.Id == 0`.  The latter would still blow if id is `nil` so a check for billingItem.Id != nil is more appropriate I guess.  How about trying it by cancelling the device to see how it behaves.